common/quotas/global/rpc/client.go

```diff
@@ -0,0 +1,191 @@
+	}
+
+	client struct {
+		history  history.Client
+		resolver history.PeerResolver
+		thisHost string
+
+		logger log.Logger
+		scope  metrics.Scope
+	}
+)
+
+var _ Client = (*client)(nil)
+
+func New(
+	historyClient history.Client,
+	resolver history.PeerResolver,
+	logger log.Logger,
+	met metrics.Client,
+) Client {
+	return &client{
+		history:  historyClient,
+		resolver: resolver,
+		thisHost: uuid.NewString(), // TODO: would descriptive be better?  but it works, unique ensures correctness.
+		logger:   logger,
+		scope:    met.Scope(metrics.FrontendGlobalRatelimiter),
+	}
+}
+
+func (c *client) Update(ctx context.Context, period time.Duration, load map[string]Calls) UpdateResult {
+	batches, err := c.resolver.GlobalRatelimitPeers(maps.Keys(load))
+	if err != nil {
+		// should only happen if peers are unavailable, individual requests are handled other ways
+		return UpdateResult{
+			Err: fmt.Errorf("unable to shard ratelimit update data: %w", err),
+		}
+	}
+
+	var mut sync.Mutex
+	weights := make(map[string]float64, len(load)) // should get back most or all keys requested
+
+	var g errgroup.Group
+	// could limit max concurrency easily with `g.SetLimit(n)` if desired,
+	// but as each goes to a different host this seems fine to just blast out all at once,
+	// and it makes timeouts easy because we don't need to reserve room for queued calls.
+	for peerAddress, batch := range batches {
+		peerAddress, batch := peerAddress, batch // for closure
+		g.Go(func() (err error) {
+			defer func() { log.CapturePanic(recover(), c.logger, &err) }()
+
+			result, err := c.updateSinglePeer(ctx, peerAddress, period, filterKeys(batch, load))
+			if err != nil {
+				return err // does not stop other calls, nor should it
+			}
+
+			mut.Lock()
+			defer mut.Unlock()
+			for k, v := range result {
+				weights[k] = v
+			}
+			return nil
+		})
+	}
+
+	err = g.Wait()
+	if err != nil {
+		// wrap it so it's relatively severe-looking, as these should not happen
+		err = fmt.Errorf("potentially fatal error during ratelimit update requests: %w", err)
+	}
+	return UpdateResult{
+		Weights: weights,
+		Err:     err,
+	}
+}
+
+func (c *client) updateSinglePeer(ctx context.Context, peerAddress string, period time.Duration, load map[string]Calls) (map[string]float64, error) {
+	anyValue, err := updateToAny(c.thisHost, period, load)
+	if err != nil {
+		// serialization errors should never happen
+		return nil, &SerializationError{err}
+	}
+
+	result, err := c.history.RatelimitUpdate(
+		ctx,
+		&types.RatelimitUpdateRequest{Any: anyValue},
+		yarpc.WithShardKey(peerAddress),
+	)
+	if err != nil {
+		// client metrics are fine for monitoring, but it does not log errors.
+		// TODO: possibly filter out or aggregate "peer lost" logs?  they're expected during deploys, since calls are not implicitly retried.
+		c.logger.Warn("request failure when updating ratelimits", tag.Error(&RPCError{err}))
```

Comment:
I would consider adding the RPC address so that we can track down a persistently problematic host